Mathlib/Control/Monad/Free.lean

```diff
@@ -0,0 +1,577 @@
+Haskell implementation that inspired this approach.
+
+## Implementation Notes
+
+The `FreeM` monad is defined using an inductive type with constructors `pure` and `liftBind`.
+We implement `Functor` and `Monad` instances, and prove the corresponding `LawfulFunctor`
+and `LawfulMonad` instances.
+
+The monads `FreeState`, `FreeWriter`, and `FreeCont` are built by supplying appropriate
+effect type constructors to the `FreeM` constructor. They are equipped with interpreters
+and helper functions.
+
+## References
+
+* [Kiselyov2015] Oleg Kiselyov, Hiromi Ishii. *Freer Monads, More Extensible Effects*.
+  Haskell Symposium 2015.
+* [MilewskiDao] Bartosz Milewski. *The Dao of Functional Programming*. 2025.
+
+## Tags
+
+Free monad, state monad, writer monad, continuation monad
+-/
+
+/-- The Free monad over a type constructor `f`.
+
+A `FreeM f a` is a tree of operations from the type constructor `f`, with leaves of type `a`.
+It has two constructors: `pure` for wrapping a value of type `a`, and `liftBind` for
+representing an operation from `f` followed by a continuation.
+
+This construction provides a free monad for any type constructor `f`, allowing for composable
+effect descriptions that can be interpreted later. Unlike the traditional free monad,
+this does not require `f` to be a functor. -/
+inductive FreeM.{u, v, w} (f : Type u → Type v) (α : Type w) where
+  | protected pure : α → FreeM f α
+  | liftBind {ι : Type u} (op : f ι) (cont : ι → FreeM f α) : FreeM f α
+
+/-
+Disable simpNF lints for auto-generated constructor lemmas, as they don't follow simp normal
+form patterns. The LHS of these lemmas use `FreeM.pure` which simplifies to `pure` via
+`pure_eq_pure`.
+-/
+attribute [nolint simpNF] FreeM.pure.sizeOf_spec FreeM.pure.injEq
+
+universe u v w
+
+namespace FreeM
+
+/-- Map a function over a `FreeM` monad. -/
+def map {α β : Type w} {F : Type u → Type v} (f : α → β) : FreeM F α → FreeM F β
+  | .pure a => .pure (f a)
+  | .liftBind op cont => .liftBind op (fun z => FreeM.map f (cont z))
+
+instance {F : Type u → Type v} : Functor (FreeM F) where
+  map := .map
+
+instance {F : Type u → Type v} : LawfulFunctor (FreeM F) where
+  map_const := rfl
+  id_map x := by
+    simp [Functor.map]
+    induction x with
+    | pure a => rfl
+    | liftBind op cont ih =>
+      simp [map, ih]
+  comp_map g h x := by
+    simp [Functor.map]
+    induction x with
+    | pure a => rfl
+    | liftBind op cont ih => simp [FreeM.map, ih]
+
+/-- Bind operation for the `FreeM` monad. -/
+protected def bind {a b : Type w} {F : Type u → Type v} (x : FreeM F a) (f : a → FreeM F b) :
+  FreeM F b :=
+  match x with
+  | .pure a => f a
+  | .liftBind op cont => .liftBind op (fun z => FreeM.bind  (cont z) f)
+
+/-- Lift an operation from the effect signature `f` into the `FreeM f` monad. -/
+@[simp]
+def lift {F : Type u → Type v} {ι : Type u} (op : F ι) : FreeM F ι :=
+  .liftBind op FreeM.pure
+
+instance {F : Type u → Type v} : Monad (FreeM F) where
```

Comment:
Can you add `variable {F : Type u → Type v}` to avoid repeating this every time?